File: examples/cpp/jobshop_sat.cc
// Copyright 2010-2025 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <algorithm>
#include <cstdint>
#include <cstdlib>
#include <string>
#include <vector>

#include "absl/flags/flag.h"
#include "absl/log/check.h"
#include "absl/log/globals.h"
#include "absl/log/log.h"
#include "examples/cpp/jobshop_sat_model.h"
#include "google/protobuf/text_format.h"
#include "google/protobuf/wrappers.pb.h"
#include "ortools/base/init_google.h"
#include "ortools/base/log_severity.h"
#include "ortools/base/types.h"
#include "ortools/sat/cp_model.h"
#include "ortools/sat/cp_model.pb.h"
#include "ortools/sat/cp_model_solver.h"
#include "ortools/sat/sat_parameters.pb.h"
#include "ortools/scheduling/jobshop_scheduling.pb.h"
#include "ortools/scheduling/jobshop_scheduling_parser.h"

ABSL_FLAG(std::string, input, "", "Jobshop data file name.");
ABSL_FLAG(std::string, params, "", "Sat parameters in text proto format.");
ABSL_FLAG(bool, use_optional_variables, false,
          "Whether we use optional variables for bounds of an optional "
          "interval or not.");
ABSL_FLAG(bool, use_interval_makespan, false,
          "Whether we encode the makespan using an interval or not.");
ABSL_FLAG(bool, use_variable_duration_to_encode_transition, true,
          "Whether we move the transition cost to the alternative duration.");
ABSL_FLAG(
    bool, use_cumulative_relaxation, true,
    "Whether we regroup multiple machines to create a cumulative relaxation.");
ABSL_FLAG(bool, display_model, false, "Display jobshop proto before solving.");
ABSL_FLAG(bool, display_sat_model, false, "Display sat proto before solving.");
ABSL_FLAG(int64_t, horizon, -1, "Override horizon computation.");

using operations_research::scheduling::jssp::Job;
using operations_research::scheduling::jssp::JobPrecedence;
using operations_research::scheduling::jssp::JsspInputProblem;
using operations_research::scheduling::jssp::Machine;
using operations_research::scheduling::jssp::Task;
using operations_research::scheduling::jssp::TransitionTimeMatrix;

namespace operations_research {
namespace sat {

// Solve a JobShop scheduling problem using CP-SAT.
void Solve(const JsspInputProblem& problem) {
  if (absl::GetFlag(FLAGS_display_model)) {
    LOG(INFO) << problem;
  }

  JobshopModelOptions options;
  options.use_optional_variables = absl::GetFlag(FLAGS_use_optional_variables);
  options.use_interval_makespan = absl::GetFlag(FLAGS_use_interval_makespan);
  options.use_variable_duration_to_encode_transition =
      absl::GetFlag(FLAGS_use_variable_duration_to_encode_transition);
  options.use_cumulative_relaxation =
      absl::GetFlag(FLAGS_use_cumulative_relaxation);
  options.horizon = absl::GetFlag(FLAGS_horizon);

  CpModelBuilder cp_model;
  const JobshopModelMapping model =
      CreateJobshopModel(problem, options, &cp_model);

  if (absl::GetFlag(FLAGS_display_sat_model)) {
    LOG(INFO) << cp_model.Proto();
  }

  // Setup parameters.
  SatParameters parameters;
  parameters.set_log_search_progress(true);

  // Tells the solver we have a makespan objective.
  // Also take decision based on precedence, this usually work better.
  parameters.set_push_all_tasks_toward_start(true);
  parameters.set_use_dynamic_precedence_in_disjunctive(true);

  // Parse the --params flag.
  if (!absl::GetFlag(FLAGS_params).empty()) {
    CHECK(google::protobuf::TextFormat::MergeFromString(
        absl::GetFlag(FLAGS_params), &parameters))
        << absl::GetFlag(FLAGS_params);
  }

  // A good default set of workers as of Jul 2026.
  if (parameters.subsolvers().empty() && parameters.num_workers() == 16) {
    LOG(INFO) << "Overriding the set of subsolvers for 16 workers.";
    parameters.add_subsolvers("fixed_no_lp");
    parameters.add_subsolvers("max_lp");
    parameters.add_subsolvers("no_lp");
    parameters.add_subsolvers("objective_shaving_no_lp");
    parameters.add_subsolvers("objective_shaving_max_lp");
    parameters.add_subsolvers("probing_no_lp");
    parameters.add_subsolvers("pseudo_costs");
    parameters.add_subsolvers("quick_restart_no_lp");
    parameters.add_subsolvers("reduced_costs");
    parameters.add_subsolvers("shaving_no_lp");
    parameters.add_subsolvers("shaving_max_lp");
  }

  const CpSolverResponse response =
      SolveWithParameters(cp_model.Build(), parameters);

  // Abort if we don't have any solution.
  if (response.status() != CpSolverStatus::OPTIMAL &&
      response.status() != CpSolverStatus::FEASIBLE)
    return;

  const int num_jobs = problem.jobs_size();

  // Check transitions.
  {
    const int num_machines = problem.machines_size();
    for (int m = 0; m < num_machines; ++m) {
      if (!problem.machines(m).has_transition_time_matrix()) continue;

      struct Data {
        int job;
        int64_t fixed_duration;
        int64_t start;
        int64_t end;
      };
      std::vector<Data> schedule;
      for (const MachineTaskData& d : model.machine_to_tasks[m]) {
        if (!SolutionBooleanValue(response, d.interval.PresenceBoolVar())) {
          continue;
        }
        schedule.push_back(
            {d.job, d.fixed_duration,
             SolutionIntegerValue(response, d.interval.StartExpr()),
             SolutionIntegerValue(response, d.interval.EndExpr())});
      }
      std::sort(schedule.begin(), schedule.end(),
                [](const Data& a, const Data& b) { return a.start < b.start; });

      const TransitionTimeMatrix& transitions =
          problem.machines(m).transition_time_matrix();

      int last_job = -1;
      int64_t last_start = kint64min;
      int64_t last_duration;
      for (const Data& d : schedule) {
        const int64_t transition =
            last_job == -1
                ? 0
                : transitions.transition_time(last_job * num_jobs + d.job);
        CHECK_LE(last_start + last_duration + transition, d.start);
        last_job = d.job;
        last_start = d.start;
        last_duration = d.fixed_duration;
      }
    }
  }

  // Check cost, recompute it from scratch.
  int64_t final_cost = 0;
  if (problem.makespan_cost_per_time_unit() != 0) {
    int64_t makespan = 0;
    for (const std::vector<JobTaskData>& tasks : model.job_to_tasks) {
      const LinearExpr job_end = tasks.back().end;
      makespan = std::max(makespan, SolutionIntegerValue(response, job_end));
    }
    final_cost += makespan * problem.makespan_cost_per_time_unit();
  }

  for (int j = 0; j < num_jobs; ++j) {
    const int64_t early_due_date = problem.jobs(j).early_due_date();
    const int64_t late_due_date = problem.jobs(j).late_due_date();
    const int64_t early_penalty =
        problem.jobs(j).earliness_cost_per_time_unit();
    const int64_t late_penalty = problem.jobs(j).lateness_cost_per_time_unit();
    const int64_t end =
        SolutionIntegerValue(response, model.job_to_tasks[j].back().end);
    if (end < early_due_date && early_penalty != 0) {
      final_cost += (early_due_date - end) * early_penalty;
    }
    if (end > late_due_date && late_penalty != 0) {
      final_cost += (end - late_due_date) * late_penalty;
    }
  }

  // Note that this the objective is a variable of the model, there is actually
  // no strong guarantee that in an intermediate solution, it is packed to its
  // minimum possible value. We do observe this from time to time. The DCHECK is
  // mainly to warn when this happen.
  //
  // TODO(user): Support alternative cost in check.
  const double tolerance = 1e-6;
  DCHECK_GE(response.objective_value(), final_cost - tolerance);
  DCHECK_LE(response.objective_value(), final_cost + tolerance);
}

}  // namespace sat
}  // namespace operations_research

int main(int argc, char** argv) {
  absl::SetStderrThreshold(absl::LogSeverityAtLeast::kInfo);
  InitGoogle(argv[0], &argc, &argv, true);

  if (absl::GetFlag(FLAGS_input).empty()) {
    LOG(FATAL) << "Please supply a data file with --input=";
  }

  operations_research::scheduling::jssp::JsspParser parser;
  CHECK(parser.ParseFile(absl::GetFlag(FLAGS_input)));
  operations_research::sat::Solve(parser.problem());
  return EXIT_SUCCESS;
}

File: examples/dotnet/JobshopSat.cs
// Copyright 2010-2021 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections.Generic;
using System.Linq;
using Google.OrTools.Sat;

class Task
{
    public Task(int taskId, int jobId, int duration, int machine)
    {
        TaskId = taskId;
        JobId = jobId;
        Duration = duration;
        Machine = machine;
        Name = "T" + taskId + "J" + jobId + "M" + machine + "D" + duration;
    }
    public int TaskId { get; set; }
    public int JobId { get; set; }
    public int Machine { get; set; }
    public int Duration { get; set; }
    public string Name { get; }
}

class JobshopSat
{
    // Number of machines.
    public const int machinesCount = 3;
    // horizon is the upper bound of the start time of all tasks.
    public const int horizon = 300;
    // this will be set to the size of myJobList variable.
    public static int jobsCount;
    /*Search time limit in milliseconds. if it's equal to 0,
    then no time limit will be used.*/
    public const int timeLimitInSeconds = 0;
    public static List<List<Task>> myJobList = new List<List<Task>>();
    public static void InitTaskList()
    {
        List<Task> taskList = new List<Task>();
        taskList.Add(new Task(0, 0, 65, 0));
        taskList.Add(new Task(1, 0, 5, 1));
        taskList.Add(new Task(2, 0, 15, 2));
        myJobList.Add(taskList);

        taskList = new List<Task>();
        taskList.Add(new Task(0, 1, 15, 0));
        taskList.Add(new Task(1, 1, 25, 1));
        taskList.Add(new Task(2, 1, 10, 2));
        myJobList.Add(taskList);

        taskList = new List<Task>();
        taskList.Add(new Task(0, 2, 25, 0));
        taskList.Add(new Task(1, 2, 30, 1));
        taskList.Add(new Task(2, 2, 40, 2));
        myJobList.Add(taskList);

        taskList = new List<Task>();
        taskList.Add(new Task(0, 3, 20, 0));
        taskList.Add(new Task(1, 3, 35, 1));
        taskList.Add(new Task(2, 3, 10, 2));
        myJobList.Add(taskList);

        taskList = new List<Task>();
        taskList.Add(new Task(0, 4, 15, 0));
        taskList.Add(new Task(1, 4, 25, 1));
        taskList.Add(new Task(2, 4, 10, 2));
        myJobList.Add(taskList);

        taskList = new List<Task>();
        taskList.Add(new Task(0, 5, 25, 0));
        taskList.Add(new Task(1, 5, 30, 1));
        taskList.Add(new Task(2, 5, 40, 2));
        myJobList.Add(taskList);

        taskList = new List<Task>();
        taskList.Add(new Task(0, 6, 20, 0));
        taskList.Add(new Task(1, 6, 35, 1));
        taskList.Add(new Task(2, 6, 10, 2));
        myJobList.Add(taskList);

        taskList = new List<Task>();
        taskList.Add(new Task(0, 7, 10, 0));
        taskList.Add(new Task(1, 7, 15, 1));
        taskList.Add(new Task(2, 7, 50, 2));
        myJobList.Add(taskList);

        taskList = new List<Task>();
        taskList.Add(new Task(0, 8, 50, 0));
        taskList.Add(new Task(1, 8, 10, 1));
        taskList.Add(new Task(2, 8, 20, 2));
        myJobList.Add(taskList);

        jobsCount = myJobList.Count;
    }

    public static void Main(String[] args)
    {
        InitTaskList();
        CpModel model = new CpModel();

        // ----- Creates all intervals and integer variables -----

        // Stores all tasks attached interval variables per job.
        List<List<IntervalVar>> jobsToTasks = new List<List<IntervalVar>>(jobsCount);
        List<List<IntVar>> jobsToStarts = new List<List<IntVar>>(jobsCount);
        List<List<IntVar>> jobsToEnds = new List<List<IntVar>>(jobsCount);

        // machinesToTasks stores the same interval variables as above, but
        // grouped my machines instead of grouped by jobs.
        List<List<IntervalVar>> machinesToTasks = new List<List<IntervalVar>>(machinesCount);
        List<List<IntVar>> machinesToStarts = new List<List<IntVar>>(machinesCount);
        for (int i = 0; i < machinesCount; i++)
        {
            machinesToTasks.Add(new List<IntervalVar>());
            machinesToStarts.Add(new List<IntVar>());
        }

        // Creates all individual interval variables.
        foreach (List<Task> job in myJobList)
        {
            jobsToTasks.Add(new List<IntervalVar>());
            jobsToStarts.Add(new List<IntVar>());
            jobsToEnds.Add(new List<IntVar>());
            foreach (Task task in job)
            {
                IntVar start = model.NewIntVar(0, horizon, task.Name);
                IntVar end = model.NewIntVar(0, horizon, task.Name);
                IntervalVar oneTask = model.NewIntervalVar(start, task.Duration, end, task.Name);
                jobsToTasks[task.JobId].Add(oneTask);
                jobsToStarts[task.JobId].Add(start);
                jobsToEnds[task.JobId].Add(end);
                machinesToTasks[task.Machine].Add(oneTask);
                machinesToStarts[task.Machine].Add(start);
            }
        }

        // ----- Creates model -----

        // Creates precedences inside jobs.
        for (int j = 0; j < jobsToTasks.Count; ++j)
        {
            for (int t = 0; t < jobsToTasks[j].Count - 1; ++t)
            {
                model.Add(jobsToEnds[j][t] <= jobsToStarts[j][t + 1]);
            }
        }

        // Adds no_overkap constraints on unary resources.
        for (int machineId = 0; machineId < machinesCount; ++machineId)
        {
            model.AddNoOverlap(machinesToTasks[machineId]);
        }

        // Creates array of end_times of jobs.
        IntVar[] allEnds = new IntVar[jobsCount];
        for (int i = 0; i < jobsCount; i++)
        {
            allEnds[i] = jobsToEnds[i].Last();
        }

        // Objective: minimize the makespan (maximum end times of all tasks)
        // of the problem.
        IntVar makespan = model.NewIntVar(0, horizon, "makespan");
        model.AddMaxEquality(makespan, allEnds);
        model.Minimize(makespan);

        // Create the solver.
        CpSolver solver = new CpSolver();
        // Set the time limit.
        if (timeLimitInSeconds > 0)
        {
            solver.StringParameters = "max_time_in_seconds:" + timeLimitInSeconds;
        }
        // Solve the problem.
        CpSolverStatus status = solver.Solve(model);

        if (status == CpSolverStatus.Optimal)
        {
            Console.WriteLine("Makespan = " + solver.ObjectiveValue);
            for (int m = 0; m < machinesCount; ++m)
            {
                Console.WriteLine($"Machine {m}:");
                SortedDictionary<long, string> starts = new SortedDictionary<long, string>();
                foreach (IntVar var in machinesToStarts[m])
                {
                    starts[solver.Value(var)] = var.Name();
                }
                foreach (KeyValuePair<long, string> p in starts)
                {
                    Console.WriteLine($"  Task {p.Value} starts at {p.Key}");
                }
            }
        }
        else
        {
            Console.WriteLine("No solution found!");
        }
    }
}
